## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 515625`
Commit: `91040c64de9beaf1d732f36277fee4c02bfad0a0`

---
### `aarch64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>wine64Packages.staging</li>
    <li>wine64Packages.stagingFull</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 7 packages built:</summary>
  <ul>
    <li>wine64 (wine64Packages.full, wine64Packages.stableFull)</li>
    <li>wine64Packages.base (wine64Packages.stable)</li>
    <li>wine64Packages.minimal</li>
    <li>wine64Packages.unstable</li>
    <li>wine64Packages.unstableFull</li>
    <li>wine64Packages.wayland</li>
    <li>wine64Packages.waylandFull</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-linux`</summary>
<details>
<summary>wine64Packages.staging</summary>
<pre>aarch64-w64-mingw32-clang -c -o dlls/ddraw/aarch64-windows/light.o dlls/ddraw/light.c -Idlls/ddraw -Iinclude -Iinclude/msvcrt -D_UCRT \
  -D__WINESRC__ -D__WINE_PE_BUILD -Wall -target aarch64-windows -fuse-ld=lld --no-default-config \
  -fno-strict-aliasing -Wempty-body -Wignored-qualifiers -Winit-self \
  -Wno-microsoft-enum-forward-reference -Wtype-limits -Wunused-but-set-parameter -Wvla \
  -Wwrite-strings -Wpointer-arith -ffunction-sections -ffp-exception-behavior=maytrap -gdwarf-4 \
  -Wdeclaration-after-statement -Wstrict-prototypes -Wabsolute-value -fasync-exceptions \
  -g -O2
aarch64-w64-mingw32-clang -c -o dlls/ddraw/aarch64-windows/main.o dlls/ddraw/main.c -Idlls/ddraw -Iinclude -Iinclude/msvcrt -D_UCRT \
  -D__WINESRC__ -D__WINE_PE_BUILD -Wall -target aarch64-windows -fuse-ld=lld --no-default-config \
  -fno-strict-aliasing -Wempty-body -Wignored-qualifiers -Winit-self \
  -Wno-microsoft-enum-forward-reference -Wtype-limits -Wunused-but-set-parameter -Wvla \
  -Wwrite-strings -Wpointer-arith -ffunction-sections -ffp-exception-behavior=maytrap -gdwarf-4 \
  -Wdeclaration-after-statement -Wstrict-prototypes -Wabsolute-value -fasync-exceptions \
  -g -O2
dlls/dcomp/tests/dcomp.c:56:2: error: &quot;Unsupported architecture&quot;
   56 | #error &quot;Unsupported architecture&quot;
      |  ^
1 error generated.
make: *** [Makefile:51304: dlls/dcomp/tests/aarch64-windows/dcomp.o] Error 1
make: *** Waiting for unfinished jobs....</pre>
</details>
<details>
<summary>wine64Packages.stagingFull</summary>
<pre>  --no-default-config -fno-strict-aliasing -Wempty-body -Wignored-qualifiers -Winit-self \
  -Wno-microsoft-enum-forward-reference -Wtype-limits -Wunused-but-set-parameter -Wvla \
  -Wwrite-strings -Wpointer-arith -ffunction-sections -ffp-exception-behavior=maytrap -gdwarf-4 \
  -Wdeclaration-after-statement -Wstrict-prototypes -Wabsolute-value -fasync-exceptions \
  -g -O2
aarch64-w64-mingw32-clang -c -o dlls/dinput/tests/aarch64-windows/joystick8.o dlls/dinput/tests/joystick8.c -Idlls/dinput/tests -Iinclude -Iinclude/msvcrt \
  -D_MSVCR_VER=0 -D__WINESRC__ -D__WINE_PE_BUILD -Wall -target aarch64-windows -fuse-ld=lld \
  --no-default-config -fno-strict-aliasing -Wempty-body -Wignored-qualifiers -Winit-self \
  -Wno-microsoft-enum-forward-reference -Wtype-limits -Wunused-but-set-parameter -Wvla \
  -Wwrite-strings -Wpointer-arith -ffunction-sections -ffp-exception-behavior=maytrap -gdwarf-4 \
  -Wdeclaration-after-statement -Wstrict-prototypes -Wabsolute-value -fasync-exceptions \
  -g -O2
dlls/dcomp/tests/dcomp.c:56:2: error: &quot;Unsupported architecture&quot;
   56 | #error &quot;Unsupported architecture&quot;
      |  ^
tools/winebuild/winebuild -w --implib -o dlls/dinput8/aarch64-windows/libdinput8.a --cc-cmd=&quot;aarch64-w64-mingw32-clang&quot; -b aarch64-windows \
  --export dlls/dinput8/dinput8.spec dlls/dinput8/aarch64-windows/data_formats.o
1 error generated.
make: *** [Makefile:51304: dlls/dcomp/tests/aarch64-windows/dcomp.o] Error 1
make: *** Waiting for unfinished jobs....</pre>
</details>
</details>

--- a/pkgs/applications/emulators/wine/base.nix
+++ b/pkgs/applications/emulators/wine/base.nix
@@ -59,7 +59,7 @@
   vulkanSupport ? false,
   sdlSupport ? false,
   usbSupport ? false,
-  mingwSupport ? stdenv.hostPlatform.isDarwin,
+  mingwSupport ? stdenv.hostPlatform.isDarwin || stdenv.hostPlatform.isAarch64,
   waylandSupport ? false,
   x11Support ? false,
   ffmpegSupport ? false,
@@ -267,6 +267,16 @@ stdenv.mkDerivation (
             lib.concatMapStringsSep " " (ps: "-W ${ps}") stagingPatches.disabledPatchsets
           }
         '';
+
+    # Fix dcomp test for aarch64
+    postPatch = lib.optionalString (useStaging && stdenv.hostPlatform.isAarch64) ''
+      if [ -f dlls/dcomp/tests/dcomp.c ]; then
+        substituteInPlace dlls/dcomp/tests/dcomp.c \
+          --replace-fail \
+            '#error "Unsupported architecture"' \
+            '__asm__ __volatile__("mov %0, sp" : "=r"(stack_pointer));'
+      fi
+    '';
 
     configureFlags =
       prevConfigFlags

--- a/pkgs/applications/emulators/wine/default.nix
+++ b/pkgs/applications/emulators/wine/default.nix
@@ -12,7 +12,13 @@ args@{
   callPackage,
   darwin,
   wineRelease ? "stable",
-  wineBuild ? if stdenv.hostPlatform.system == "x86_64-linux" then "wineWow" else "wine32",
+  wineBuild ?
+    if stdenv.hostPlatform.system == "x86_64-linux" then
+      "wineWow"
+    else if stdenv.hostPlatform.isAarch64 then
+      "wine64"
+    else
+      "wine32",
   gettextSupport ? false,
   fontconfigSupport ? false,
   alsaSupport ? false,
@@ -39,7 +45,7 @@ args@{
   vulkanSupport ? false,
   sdlSupport ? false,
   usbSupport ? false,
-  mingwSupport ? stdenv.hostPlatform.isDarwin,
+  mingwSupport ? stdenv.hostPlatform.isDarwin || stdenv.hostPlatform.isAarch64,
   waylandSupport ? false,
   x11Support ? false,
   ffmpegSupport ? false,

--- a/pkgs/applications/emulators/wine/llvm-mingw.nix
+++ b/pkgs/applications/emulators/wine/llvm-mingw.nix
@@ -1,0 +1,69 @@
+{
+  lib,
+  stdenvNoCC,
+  fetchurl,
+  autoPatchelfHook,
+  zlib,
+  stdenv,
+}:
+
+# Used as a workaround for ucrtAarch64.
+# compiler-rt due to pthread.h not being found during cross-build.
+
+let
+  version = "20260421";
+
+  hostArch =
+    if stdenv.hostPlatform.isAarch64 then
+      "aarch64"
+    else if stdenv.hostPlatform.isx86_64 then
+      "x86_64"
+    else
+      throw "llvm-mingw: unsupported host platform ${stdenv.hostPlatform.system}";
+
+  hashes = {
+    aarch64 = "sha256-6c+lqTKQy4Utxaa5DM3iVUA6YzMmSL4XFHDKPBrTeCg=";
+    x86_64 = "sha256-+LjM53mv/qtHvK7Gzm6edosWYJSjZhI+9kstCzicwSE=";
+  };
+in
+stdenvNoCC.mkDerivation {
+  pname = "llvm-mingw-${hostArch}";
+  inherit version;
+
+  src = fetchurl {
+    url = "https://github.com/mstorsjo/llvm-mingw/releases/download/${version}/llvm-mingw-${version}-ucrt-ubuntu-22.04-${hostArch}.tar.xz";
+    hash = hashes.${hostArch};
+  };
+
+  nativeBuildInputs = [ autoPatchelfHook ];
+  buildInputs = [
+    (lib.getLib stdenv.cc.cc) # libstdc++
+    zlib
+  ];
+
+  dontConfigure = true;
+  dontBuild = true;
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out
+    cp -r * $out/
+
+    runHook postInstall
+  '';
+
+  autoPatchelfIgnoreMissingDeps = [ "*.dll" ];
+  dontStrip = true;
+
+  meta = {
+    description = "LLVM/Clang/LLD-based mingw-w64 toolchain (prebuilt)";
+    homepage = "https://github.com/mstorsjo/llvm-mingw";
+    license = lib.licenses.zlib;
+    platforms = [
+      "aarch64-linux"
+      "x86_64-linux"
+    ];
+    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
+  };
+}

--- a/pkgs/applications/emulators/wine/packages.nix
+++ b/pkgs/applications/emulators/wine/packages.nix
@@ -28,6 +28,8 @@ let
     mono
     ;
 
+  llvm-mingw = callPackage ./llvm-mingw.nix { };
+
   # Args to pass through to base.nix (support flags, etc.)
   baseArgs = removeAttrs args [
     "stdenv_32bit"
@@ -71,13 +73,19 @@ in
       pname = "wine64";
       inherit version patches;
       pkgArches = [ pkgs ];
-      mingwGccs = with pkgsCross; [ mingwW64.buildPackages.gcc ];
+      mingwGccs =
+        if pkgs.stdenv.hostPlatform.isAarch64 then
+          [ llvm-mingw ]
+        else
+          with pkgsCross; [ mingwW64.buildPackages.gcc ];
       geckos = [ gecko64 ];
       monos = [ mono ];
-      configureFlags = [ "--enable-win64" ];
+      configureFlags =
+        if pkgs.stdenv.hostPlatform.isAarch64 then [ "--enable-archs=aarch64" ] else [ "--enable-win64" ];
       platforms = [
         "x86_64-linux"
         "x86_64-darwin"
+        "aarch64-linux"
       ];
       mainProgram = "wine";
     }
@@ -119,16 +127,26 @@ in
       inherit version patches;
       mingwSupport = true; # Required because we request "--enable-archs=x86_64"
       pkgArches = [ pkgs ];
-      mingwGccs = with pkgsCross; [
-        mingw32.buildPackages.gcc
-        mingwW64.buildPackages.gcc
-      ];
+      mingwGccs =
+        if pkgs.stdenv.hostPlatform.isAarch64 then
+          [ llvm-mingw ]
+        else
+          with pkgsCross;
+          [
+            mingw32.buildPackages.gcc
+            mingwW64.buildPackages.gcc
+          ];
       geckos = [ gecko64 ];
       monos = [ mono ];
-      configureFlags = [ "--enable-archs=x86_64,i386" ];
+      configureFlags =
+        if pkgs.stdenv.hostPlatform.isAarch64 then
+          [ "--enable-archs=aarch64,x86_64,i386" ]
+        else
+          [ "--enable-archs=x86_64,i386" ];
       platforms = [
         "x86_64-linux"
         "x86_64-darwin"
+        "aarch64-linux"
       ];
       mainProgram = "wine";
     }

